### ISSUE TITLE
Feature request: add way to know which file a test originated from

### ISSUE THREAD

**Issue by nicojs** [jasmine/jasmine#1884]
Currently, there doesn't seem to be a way to know from which test file a test originated. I was trying to find it on the custom reporter here: https://jasmine.github.io/api/edge/Reporter.html. 

Are there ideas to add this feature? For example:

```js
jasmine.env.addReporter({
  suiteStarted(s) {
    console.log(s.fileName);
  },
  specStarted(spec) {
    console.log(spec.fileName);
  }
});
```

NB: ~~Both mocha and~~ jest supports this feature.

It would be useful for [Stryker, the JavaScript mutation testing framework](https://stryker-mutator.io), since we are in the process of adding information.

**Comment by nicojs** [comment]
Thanks for your extensive reply! Interesting stuff.

> As far as I know there's no portable API to get a machine-readable stack trace.

Yeah, no way around parsing the stack message I'm afraid. If you can detect that you're running on v8, you could use the [Error.prepareStackTrace](https://v8.dev/docs/stack-trace-api#customizing-stack-traces) API (see my workaround above), but for other engines, you will have to use the stack message itself.

I would be happy with an implementation that is best-effort. If someone used `eval` or a wrapper around `it`, they would see those locations. We could also make this optional behind a flag [like Jest has](https://jestjs.io/docs/en/cli#--testlocationinresults)

F.y.i. sorry for reporting this issue to the wrong repo 😓

**Comment by nicojs** [comment]
Let me do a proposal.

I want to create and maintain a good package that provides you with the most accurate location of the call site. I've already started it here: https://github.com/nicojs/call-id. It will support all actively used browsers (as long as sauce labs or BrowserStack allows me to test them). It can be used as an ESM or CJS module, so it should be easily bundled in browser packages.

I want to implement this feature in both mocha (https://github.com/mochajs/mocha/issues/4572) and here in jasmine. What do you think?

**Comment by elliot-nelson** [comment]
Another possible option, that might be easier for the test libraries to swallow, is if it was just a plugin instead. That you if you want it, you can enable it like any other plugin/matcher (in `helpers.js` for example).

You could patch in additional information to the test results object itself, or use it as a wrapper e.g.:

```js
// helpers.js

const callId = require('call-id');

jasmine.env.addReporter({
  suiteStarted(s) {
    console.log(callId(s).fileName);
  },
  specStarted(spec) {
    console.log(callId(spec).fileName);
  }
});
```

**Comment by sgravrock** [comment]
@nicojs I appreciate you taking the initiative to tackle this. Doing the stack trace processing in a library makes a lot of sense in principle. But Jasmine isn't currently set up to depend on external libraries, and there are [some problems](https://github.com/jasmine/jasmine/issues/1800#issuecomment-598982016) to solve before we can use tools like Babel or Webpack or depend on libraries that are built with those tools.

Because of that, I think we're better off using (and if necessary, improving) the stack trace reading code that we already have. I'd be happy to review a PR that does that, provided that it works with our currently supported browsers & Node versions and fails in a sane way if some future environment doesn't meet its expectations.

**Comment by nicojs** [comment]
> Another possible option, that might be easier for the test libraries to swallow, is if it was just a plugin instead. 

That would be fine for me as well, the problem is that [`call-id`](https://github.com/nicojs/call-id#-example) needs to know the exact distance on the stack. In other words, the internals of jasmine should be aligned with the plugin. So if the distance is `3` today, it should remain 3, regardless of internal restructuring. In other words: this would limit the maintainers of this jasmine in any future refactoring work. It would be fine for me (the plugin creator), but then I would like to add integration tests that verify this behavior keeps working.

@elliot-nelson I see in your API proposal, you provide a function to `callId`: `suiteStarted(s) { console.log(callId(s).fileName); }`. That would only work in the v8- call stack API. There is no SpiderMonkey implementation that does something equivalent, and thus we should rely on stack parsing, unfortunately 🤷‍♂️.

> Doing the stack trace processing in a library makes a lot of sense in principle. But Jasmine isn't currently set up to depend on external libraries, and there are some problems to solve before we can use tools like Babel or Webpack or depend on libraries that are built with those tools.

If you would depend on `call-id`, you would not have to compile typescript or something like it. It comes pre-compiled to JS (type definitions are included, but you don't have to use them of course) (requiring dependents of your library to compile your code is a definite bad practice). I currently provide a hybrid package (commonjs and esm) as well as it provides a browser bundle, so maximum flexibility.

> I'd be happy to review a PR that does that, provided that it works with our currently supported browsers & Node versions and fails in a sane way if some future environment doesn't meet its expectations.

In Mocha source code, we allow dependencies, but they are bundled in for the browser bundle. I can do something similar here.

**Comment by ajvincent** [comment]
If you write your tests as ES6 modules, there's usually `import.meta.url`.  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.meta

**Comment by sgravrock** [comment]
Most spec files aren't ES modules, and I don't expect that to change for the foreseeable future. However, while we can't assume ES modules, we *can* make a lot more assumptions about `Error` objects now that we've dropped support for some of the more problematic platforms. I put together a quick spike in 805288c1. That spec passes on all supported browsers and Node versions.

@nicojs What do you think? Would something like that work well enough? The obvious gotchas I can think of are cases where the actual `it` call is in a library or wrapped in `eval`, and situations where the spec file is compiler/bundler output and the runtime doesn't apply source maps to stack traces. But I think any approach is going to break down in those cases.

**Comment by sgravrock** [comment]
@nicojs There's been enough interest in this feature from other people that I'm going to go ahead and ship 6ad8d20694fc4bd995b88e21addd2c35ccbc9bb0 in the next release if I don't hear any objections. Consider this the last call for feedback.

**Comment by dmrickey** [comment]
@sgravrock 
![Image](https://github.com/user-attachments/assets/14eff2e7-322e-466d-be99-f7782aa9a48b)

I'm confused with what I'm seeing. I expected the local file path but that is not the case. I'm not even sure how what it's giving me is supposed to be useful

**Comment by sgravrock** [comment]
This is just a guess, but if you have an Angular project you might be running into a known bug in zone.js: https://github.com/jasmine/jasmine/issues/2016

If that's not it, there's not nearly enough for me to go on. (This is one reason why so many projects auto-lock old issues: "this doesn't work for me" comments on old resolved issues almost never have enough information for somebody else to investigate.) If you're able to reproduce the problem without zone.js or anything else that monkey patches Jasmine, please feel to open a new issue with complete, working steps to reproduce.
